text/0000-deprecate-string-based-actions.md

```diff
@@ -0,0 +1,89 @@
+- Start Date: 2020-05-23
+- Relevant Team(s): Ember.js
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Deprecate String Based Actions
+
+## Motivation
+
+> In Ember 1.13 we introduced [closure actions](https://github.com/emberjs/rfcs/blob/00ac2685c86f27d41547012903f485a4ef338d27/active/0000-improved-actions.md) have been recommended over string based actions. This officially deprecates string based actions in favor of closure actions,
+which provide many benefits such as
+
+1. simpler passing through components
+2. less need for an actions hash
+3. ability to return a value
+4. typing in TypeScript
+5. compatible with Glimmer Components
+
+## Detailed Design
+
+We will deprecate the following:
+
+1. The `send` api.
+2. The use of a string parameter to the `action` helper.
+
+## Transition Path
+
+Users will transition to closure actions.
+
+From:
+
+```js
+export default Component.extend({
+    someMethod() {
+        this.send('actionName');
+    }
+
+    actions: {
+        actionName() {
+            // do something
+        }
+    }
+});
+```
+
+To:
+
+```js
+export default Component.extend({
+    someMethod() {
+        this.actionName();
+    }
+
+    actionName() {
+        // do something
+    }
+});
+```
+
+From:
+
+```hbs
+<button {{action 'actionName'}}>Action Label</button>
+```
+
+To:
+
+```hbs
+<button {{action this.actionName}}>Action Label</button>
```

Comment:
Honestly, we should not suggest using `action` here at all. While this RFC doesn't explicitly deprecate `{{action` / `(action` we shouldn't suggest folks migrate from `{{action 'foo'}}` to `{{action this.foo}}` when we _really think_ they ought to be using `{{on 'click' this.foo}}`. Doing that migration is just forcing double work / churn on folks.

Reply:
Are we planning to deprecate the `action` helper entirely in the near future?

Reply:
I was thinking it might be easier not to encourage people to go all the way to Octane recommendations, not to mention easier to write a codemod that just converted string based actions to closure actions?

Reply:
Neither helper nor modifier form of `{{action` are part of the octane programming model. Therefore, it seems very likely that both forms will be fully deprecated at some point in the future. 

IMHO, the only question here is what "soon" actually means. It seems unlikely that `{{action` or `(action` would be something that we would attempt to _remove_ in a 4.0 (it will take time for applications to migrate), but I would expect them to have been deprecated before then.

Reply:
The issue is, rewriting from one deprecated thing (deprecated in this RFC) to another thing that is likely to be deprecated soon (when there is a perfectly good, and easier to understand alternative that will not be deprecated) is just introducing churn for no good reason.

Reply:
> IMHO, the only question here is what "soon" actually means.

Only if we intend to deprecate the action helper and modifier prior to deprecating Ember Components, Controllers, and Ember.Object, would the user really experience extra churn. However, if the action helper and modifier will exist until these are deprecated, then no extra churn will be created and there is no need to immediately stop using `action` modifier and helper altogether.

Reply:
Does an an alternative need to be detailed at all here?

Reply:
@mehulkar I think an alternative should be detailed here, that is why I did so. It **definitely** needs to be detailed in the deprecation guide.

Reply:
Maybe this RFC / deprecation guide can say something like:

> While the action helper and modifiers are still supported with non-string arguments, the refactoring recommendation is to switch to the `on` modifier and `@action` decorator directly. It is likely that the `action` helpers and modifiers will be deprecated in future RFCs.

and give an example like this:

> For example, take this button that sends the `foo` action with the default click event:

> ```hbs
> <button {{action "foo"}}></button>
> ```

> Although this _could_ be refactored to:

> ```hbs
> <button {{action this.foo}}></button>
> ```

> where `foo()` is implemented in the backing context in the actions hash:

> ```js
> Component.extend({
>   actions: {
>     foo() {}
>   }
> })
> ```

> it is recommended that the code is refactored instead to:

> ```hbs
> <button {{on "click" this.foo}}></button>
> ```

> ```js
> Component.extend({
>   foo: action(function() {})
> })
> ```

This could also be written as the with the `diff` syntax to make it more clear what changed. It doesn't quite capture the case of refactoring away from an action that bubbles, but you could link to a page in the guides that covers it.

Reply:
This seems good. :+1: